File: 06/gpt-5.1/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func solve(lines []string) int64 {
	if len(lines) == 0 {
		return 0
	}

	height := len(lines)
	width := 0
	for _, line := range lines {
		if len(line) > width {
			width = len(line)
		}
	}

	// Pad lines to equal width
	grid := make([]string, height)
	for i, line := range lines {
		trimmed := strings.TrimRight(line, "\r\n")
		if len(trimmed) < width {
			trimmed = trimmed + strings.Repeat(" ", width-len(trimmed))
		}
		grid[i] = trimmed
	}

	type rng struct{ start, end int }
	ranges := []rng{}
	inBlock := false
	startCol := 0

	for x := 0; x < width; x++ {
		allSpace := true
		for y := 0; y < height; y++ {
			if grid[y][x] != ' ' {
				allSpace = false
				break
			}
		}

		if allSpace {
			if inBlock {
				ranges = append(ranges, rng{start: startCol, end: x - 1})
				inBlock = false
			}
		} else if !inBlock {
			inBlock = true
			startCol = x
		}
	}

	if inBlock {
		ranges = append(ranges, rng{start: startCol, end: width - 1})
	}

	var total int64 = 0

	for _, r := range ranges {
		start, end := r.start, r.end
		// Find operator in bottom row
		var op byte = 0
		for x := start; x <= end; x++ {
			c := grid[height-1][x]
			if c == '+' || c == '*' {
				op = c
				break
			}
		}
		if op == 0 {
			continue
		}

		nums := []int64{}
		for y := 0; y < height-1; y++ {
			segment := grid[y][start : end+1]
			s := strings.TrimSpace(segment)
			if s == "" {
				continue
			}
			var n int64
			fmt.Sscanf(s, "%d", &n)
			nums = append(nums, n)
		}

		if len(nums) == 0 {
			continue
		}

		var value int64
		if op == '+' {
			value = 0
			for _, n := range nums {
				value += n
			}
		} else { // '*'
			value = 1
			for _, n := range nums {
				value *= n
			}
		}

		total += value
	}

	return total
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	lines := []string{}
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	// Part 1
	part1 := solve(lines)
	fmt.Printf("Day 06 Part 1: %d\n", part1)

	// Part 2
	part2 := solvePart2(lines)
	fmt.Printf("Day 06 Part 2: %d\n", part2)
}

func solvePart2(lines []string) int64 {
	if len(lines) == 0 {
		return 0
	}

	height := len(lines)
	width := 0
	for _, line := range lines {
		if len(line) > width {
			width = len(line)
		}
	}

	grid := make([]string, height)
	for i, line := range lines {
		trimmed := strings.TrimRight(line, "\r\n")
		if len(trimmed) < width {
			trimmed = trimmed + strings.Repeat(" ", width-len(trimmed))
		}
		grid[i] = trimmed
	}

	type rng struct{ start, end int }
	ranges := []rng{}
	inBlock := false
	startCol := 0

	for x := 0; x < width; x++ {
		allSpace := true
		for y := 0; y < height; y++ {
			if grid[y][x] != ' ' {
				allSpace = false
				break
			}
		}

		if allSpace {
			if inBlock {
				ranges = append(ranges, rng{start: startCol, end: x - 1})
				inBlock = false
			}
		} else if !inBlock {
			inBlock = true
			startCol = x
		}
	}

	if inBlock {
		ranges = append(ranges, rng{start: startCol, end: width - 1})
	}

	var total int64 = 0

	for _, r := range ranges {
		start, end := r.start, r.end

		var op byte = 0
		for x := start; x <= end; x++ {
			c := grid[height-1][x]
			if c == '+' || c == '*' {
				op = c
				break
			}
		}
		if op == 0 {
			continue
		}

		nums := []int64{}

		for x := end; x >= start; x-- {
			var digits strings.Builder
			for y := 0; y < height-1; y++ {
				c := grid[y][x]
				if c >= '0' && c <= '9' {
					digits.WriteByte(c)
				}
			}

			if digits.Len() > 0 {
				var n int64
				fmt.Sscanf(digits.String(), "%d", &n)
				nums = append(nums, n)
			}
		}

		if len(nums) == 0 {
			continue
		}

		var value int64
		if op == '+' {
			value = 0
			for _, n := range nums {
				value += n
			}
		} else {
			value = 1
			for _, n := range nums {
				value *= n
			}
		}

		total += value
	}

	return total
}

File: 06/gpt-5.1/rust/src/main.rs
use std::fs;

fn solve(lines: &[String]) -> i64 {
    if lines.is_empty() {
        return 0;
    }

    let height = lines.len();
    let mut width = 0usize;
    for line in lines {
        if line.len() > width {
            width = line.len();
        }
    }

    // Pad lines to equal width
    let mut grid: Vec<Vec<char>> = Vec::with_capacity(height);
    for line in lines {
        let mut chars: Vec<char> = line.trim_end_matches(['\r', '\n']).chars().collect();
        if chars.len() < width {
            chars.extend(std::iter::repeat(' ').take(width - chars.len()));
        }
        grid.push(chars);
    }

    #[derive(Copy, Clone)]
    struct Range {
        start: usize,
        end: usize,
    }

    let mut ranges: Vec<Range> = Vec::new();
    let mut in_block = false;
    let mut start_col = 0usize;

    for x in 0..width {
        let mut all_space = true;
        for y in 0..height {
            if grid[y][x] != ' ' {
                all_space = false;
                break;
            }
        }

        if all_space {
            if in_block {
                ranges.push(Range { start: start_col, end: x - 1 });
                in_block = false;
            }
        } else if !in_block {
            in_block = true;
            start_col = x;
        }
    }

    if in_block {
        ranges.push(Range { start: start_col, end: width - 1 });
    }

    let mut total: i64 = 0;

    for range in ranges {
        let start = range.start;
        let end = range.end;

        // Find operator in bottom row
        let mut op: Option<char> = None;
        for x in start..=end {
            let c = grid[height - 1][x];
            if c == '+' || c == '*' {
                op = Some(c);
                break;
            }
        }
        let op = match op {
            Some(c) => c,
            None => continue,
        };

        let mut nums: Vec<i64> = Vec::new();
        for y in 0..(height - 1) {
            let segment: String = grid[y][start..=end].iter().collect();
            let s = segment.trim();
            if s.is_empty() {
                continue;
            }
            let n: i64 = s.parse().expect("invalid number");
            nums.push(n);
        }

        if nums.is_empty() {
            continue;
        }

        let value: i64 = if op == '+' {
            nums.iter().sum()
        } else {
            nums.iter().product()
        };

        total += value;
    }

    total
}

fn main() {
    let raw = fs::read_to_string("input.txt").expect("Failed to read input");
    let lines: Vec<String> = raw.lines().map(|s| s.to_string()).collect();

    // Part 1
    let part1: i64 = solve(&lines);
    println!("Day 06 Part 1: {}", part1);

    // Part 2
    let part2: i64 = solve_part2(&lines);
    println!("Day 06 Part 2: {}", part2);
}

fn solve_part2(lines: &[String]) -> i64 {
    if lines.is_empty() {
        return 0;
    }

    let height = lines.len();
    let mut width = 0usize;
    for line in lines {
        if line.len() > width {
            width = line.len();
        }
    }

    let mut grid: Vec<Vec<char>> = Vec::with_capacity(height);
    for line in lines {
        let mut chars: Vec<char> = line.trim_end_matches(['\r', '\n']).chars().collect();
        if chars.len() < width {
            chars.extend(std::iter::repeat(' ').take(width - chars.len()));
        }
        grid.push(chars);
    }

    #[derive(Copy, Clone)]
    struct Range {
        start: usize,
        end: usize,
    }

    let mut ranges: Vec<Range> = Vec::new();
    let mut in_block = false;
    let mut start_col = 0usize;

    for x in 0..width {
        let mut all_space = true;
        for y in 0..height {
            if grid[y][x] != ' ' {
                all_space = false;
                break;
            }
        }

        if all_space {
            if in_block {
                ranges.push(Range { start: start_col, end: x - 1 });
                in_block = false;
            }
        } else if !in_block {
            in_block = true;
            start_col = x;
        }
    }

    if in_block {
        ranges.push(Range { start: start_col, end: width - 1 });
    }

    let mut total: i64 = 0;

    for range in ranges {
        let start = range.start;
        let end = range.end;

        let mut op: Option<char> = None;
        for x in start..=end {
            let c = grid[height - 1][x];
            if c == '+' || c == '*' {
                op = Some(c);
                break;
            }
        }
        let op = match op {
            Some(c) => c,
            None => continue,
        };

        let mut nums: Vec<i64> = Vec::new();

        for x in (start..=end).rev() {
            let mut digits = String::new();
            for y in 0..(height - 1) {
                let c = grid[y][x];
                if c.is_ascii_digit() {
                    digits.push(c);
                }
            }

            if !digits.is_empty() {
                let n: i64 = digits.parse().expect("invalid number");
                nums.push(n);
            }
        }

        if nums.is_empty() {
            continue;
        }

        let value: i64 = if op == '+' {
            nums.iter().sum()
        } else {
            nums.iter().product()
        };

        total += value;
    }

    total
}
